Tasks/ServiceFabricComposeDeploy/ServiceFabricComposeDeploy.ps1

```diff
@@ -163,48 +163,101 @@ try {
-            $existingApplication = Get-ServiceFabricComposeApplicationStatusHelper -ApiVersion $apiVersion -GetStatusParameters $getStatusParameters
+            $upgrading = $true
         }
-        while ($existingApplication -ne $null)
-        Write-Host (Get-VstsLocString -Key ApplicationRemoved)
-    }
+        else
+        {
+            Write-Host (Get-VstsLocString -Key RemovingApplication -ArgumentList $applicationName)
 
-    Write-Host (Get-VstsLocString -Key CreatingApplication)
-    New-ServiceFabricComposeApplicationHelper -ApiVersion $apiVersion -DeployParameters $deployParameters
+            Remove-ServiceFabricComposeApplicationHelper -ApiVersion $apiVersion -RemoveParameters $removeParameters
+            do
+            {
+                Write-Host (Get-VstsLocString -Key CurrentStatus -ArgumentList $existingApplication.Status)
+                Start-Sleep -Seconds 3
+                $existingApplication = Get-ServiceFabricComposeApplicationStatusHelper -ApiVersion $apiVersion -GetStatusParameters $getStatusParameters
+            }
+            while ($existingApplication -ne $null)
+            Write-Host (Get-VstsLocString -Key ApplicationRemoved)
+        }
+    }
 
-    Write-Host (Get-VstsLocString -Key WaitingForDeploy)
-    $newApplication = Get-ServiceFabricComposeApplicationStatusHelper -ApiVersion $apiVersion -GetStatusParameters $getStatusParameters
-    while (($newApplication -eq $null) -or `
-           ($newApplication.Status -eq 'Provisioning') -or `
-           ($newApplication.Status -eq 'Creating'))
+    if ($upgrading)
     {
-        if ($newApplication -eq $null)
+        $upgradeParameters = New-Object 'System.Collections.Hashtable' $deployParameters
+        $upgradeParameters.Monitored = $True
+        $upgradeParameters.FailureAction = 'Rollback'
+        $upgradeParameters.ConsiderWarningAsError = $True
+        $upgradeParameters.Force = $True
+
+        $upgradeStatus = Get-ServiceFabricComposeDeploymentUpgradeHelper -ApiVersion $apiVersion -GetUpgradeParameters $getStatusParameters
+        if (($upgradeStatus -ne $null) -and (IsUpgradeRunning $upgradeStatus.UpgradeState))
         {
-            Write-Host (Get-VstsLocString -Key WaitingForDeploy)
+            Write-Error (Get-VstsLocString -Key UpgradeInProgress -ArgumentList $applicationName)
         }
-        else
+
+        Write-Host (Get-VstsLocString -Key UpgradingApplication)
+        Start-ServiceFabricComposeDeploymentUpgradeHelper -ApiVersion $apiVersion -UpgradeParameters $upgradeParameters
+
+        Write-Host (Get-VstsLocString -Key WaitingForUpgrade)
+        # Wait a minute before checking on the upgrade to avoid getting the status of the last upgrade.
+        Start-Sleep -Seconds 60
+        $upgradeStatus = Get-ServiceFabricComposeDeploymentUpgradeHelper -ApiVersion $apiVersion -GetUpgradeParameters $getStatusParameters
+        while (($upgradeStatus -eq $null) -or (IsUpgradeRunning $upgradeStatus.UpgradeState))
```

Comment:
Will `($upgradeStatus -eq $null)` be `true` if upgrade completes within 60 seconds we waited earlier - this will put us in infinite loop

Reply:
No, once the upgrade is complete the Get-ServiceFabricComposeDeploymentUpgradeHelper will continue to return the results of the last upgrade. This can only be $null when no upgrade has ever been performed on the application or within a narrow window after starting the first upgrade.